### DISCUSSION TITLE
Android system certificate invalid

### DISCUSSION THREAD

**Discussion by Gunkkk** [mitmproxy/mitmproxy#5312] [Intercept Everything!]
**platform: LineageOS 18.1**

I want to intercept the HTTPS traffic using mitmproxy, and when i install the certificate as user in Android,  it works in Chrome at least.
I hope it works for most Apps, so I install the certificete in \system\etc\security\cacerts as common, and something is wrong. 
I can check the installed system CAs in settings, and the mitmproxy certificate was installed, but the page reports that 

> NET::ERR_CERT_AUTHORITY_INVALID  or ERR_CERTIFICATE_TRANSPARENCY_REQUIRED

and mitmdump reports  

> TlsException("SSL handshake error: Error([('SSL routines', 'ssl3_read_bytes', 'sslv3 alert certificate unknown')])") 

I have no idea whether the device problem because it worked just months ago.
Thanks!

**Comment by atarjoe** [comment]
I also have this exact problem with LineageOS 14.1 and Chrome 100.0.4896.127.
Prior to this Chrome version it worked okay.
After a bit research it appears to be a new feature introduced by Google engineers (and I believe will be incorporated gradually to Chrome on PCs too) in order to help sites owners to track and check of unauthorized certificates issuance for their sites.
Read here for more info:

https://certificate.transparency.dev/

**Reply by mhils** [reply]
Does this problem only affect Chrome, or other apps as well? If it's Chrome only, could you try to install the certificate in the user store as well?

**Reply by Gunkkk** [reply]
> )

Well, it seems that the Certificate Transparency check in new version Chrome leats to that.
Is there  any other solution besides back to older Chrome since I cannot installed it as system's certificate and user's  at the same time?
And I only find the problem in Chrome until now. @mhils
Thanks!

**Reply by mhils** [reply]
> since I cannot installed it as system's certificate and user's at the same time

Can you clarify why that is not possible please?

**Reply by Gunkkk** [reply]
> > since I cannot installed it as system's certificate and user's at the same time
> 
> Can you clarify why that is not possible please?

Sorry for my subjective assertion. 
Although it cannot be installed by the UI, we could do it as [https://stackoverflow.com/a/58292316](https://stackoverflow.com/a/58292316).
Generally as below, push the certs to /data/misc/user/0/cacerts-added/ with the old hash name.
> adb push ./$subjectHash.0 /data/misc/user/0/cacerts-added/$subjectHash.0

And the problem still exists.

**Reply by mhils** [reply]
> Although it cannot be installed by the UI

Sorry for the stupid question, but why can it not be installed? Do I get you right that Android essentially blocks that because it's already present as a system cert?

**Reply by Gunkkk** [reply]
> > Although it cannot be installed by the UI
> 
> Sorry for the stupid question, but why can it not be installed? Do I get you right that Android essentially blocks that because it's already present as a system cert?

Well in fact we can install the user certs using adb and  we can check that in setting/security...
It seems that the Android will handle them independently as in [source code](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/security/net/config/XmlConfigSource.java). 
I guess that the Chrome handling logic leads to the repeat user certs useless.
Is there any possiblility for mitmproxy to solve the emerging Certificate Transparency(CT) mechanism please？

**Reply by mhils** [reply]
See https://chromium.googlesource.com/chromium/src/+/master/net/docs/certificate-transparency.md#locally_trusted-cas (emphasis mine):

> For CAs that have been manually installed, provided those certificates are not or have not been publicly-trusted, it‘s not necessary to enable support for Certificate Transparency. Further, **Certificate Transparency Logs will not accept certificates from those CAs, thus it’s not possible to support CT.**

You need to install the certificate as a user-added CA.

**Reply by Gunkkk** [reply]
> For CAs that have been manually installed, provided those certificates are not or have not been publicly-trusted, it‘s not necessary to enable support for Certificate Transparency. Further, **Certificate Transparency Logs will not accept certificates from those CAs, thus it’s not possible to support CT.**

Thanks! I will continue to explore how to setting the simultaneous certs.

**Comment by atarjoe** [comment]
Currently, it seems that it affects only on Chrome but other apps'
developers may choose to implement this new standard whixh is defined here:
https://datatracker.ietf.org/doc/html/rfc6962

MDM reported (
https://developer.mozilla.org/en-US/docs/Web/Security/Certificate_Transparency)
that Firefox currently doesn't require CT logs for sites that users visit,
but Chrome and Apple safari does require it for some sites.
